File: src/acl/grants.ts
/**
 * Least-privilege ACL engine (§5, §7).
 *
 * Default deny. Everything an agent may touch must be listed on the agent:
 * MCPs, repos, env, filesystem folders, collaboration spawns, network hosts.
 * Filesystem ops are authorized server-side per folder grant + verb.
 */
import type {
  Agent,
  FilesystemGrant,
  FileObject,
  SessionManifest,
  Skill,
} from "../domain/types.js";

export type FsOp = "list" | "read" | "write" | "mkdir" | "delete";

/** Normalize a virtual path; returns null when escaping the root. */
export function normalizePath(raw: string): string | null {
  const parts = raw.split("/");
  const out: string[] = [];
  for (const p of parts) {
    if (p === "" || p === ".") continue;
    if (p === "..") return null; // traversal is denied outright
    out.push(p);
  }
  return "/" + out.join("/");
}

export function pathWithin(prefix: string, path: string): boolean {
  const p = normalizePath(prefix);
  const q = normalizePath(path);
  if (!p || !q) return false;
  if (p === "/") return true;
  return q === p || q.startsWith(p.endsWith("/") ? p : p + "/");
}

/** Combined grant for a path across all the agent's folder grants. */
export function grantFor(grants: FilesystemGrant[], path: string): FilesystemGrant {
  const merged: FilesystemGrant = {
    folderPath: path,
    canRead: false,
    canWrite: false,
    canDelete: false,
  };
  for (const g of grants) {
    if (pathWithin(g.folderPath, path)) {
      merged.canRead ||= g.canRead;
      merged.canWrite ||= g.canWrite;
      merged.canDelete ||= g.canDelete;
    }
  }
  return merged;
}

/** §7 server-side authorization: deny unless the grant covers the op. */
export function checkFsOp(
  grants: FilesystemGrant[],
  op: FsOp,
  path: string,
): { ok: true } | { ok: false; reason: string } {
  const norm = normalizePath(path);
  if (!norm) return { ok: false, reason: `path escape denied: ${path}` };
  const g = grantFor(grants, norm);
  if (!g.canRead && (op === "list" || op === "read"))
    return { ok: false, reason: `no read grant for ${norm}` };
  if (!g.canWrite && (op === "write" || op === "mkdir"))
    return { ok: false, reason: `no write grant for ${norm}` };
  if (!g.canDelete && op === "delete")
    return { ok: false, reason: `no delete grant for ${norm}` };
  return { ok: true };
}

/** The session manifest is the ONLY envelope a runner receives (§8.1). */
export function buildSessionManifest(input: {
  projectId: string;
  sessionId: string;
  agent: Agent;
  skills: Skill[];
  envNames: string[];
  environment: SessionManifest["environment"];
  task: SessionManifest["task"];
  goal: SessionManifest["goal"];
  attachments: FileObject[];
  /** MCP connection names (agent-facing contract); defaults to agent ids. */
  mcpConnectionNames?: string[];
}): SessionManifest {
  const { agent, skills, envNames, environment, task, goal, attachments } = input;
  return {
    sessionId: input.sessionId,
    projectId: input.projectId,
    agent: {
      id: agent.id,
      name: agent.name,
      title: agent.title,
      model: agent.model,
      foundationalPrompt: agent.foundationalPrompt,
      rolePrompt: agent.rolePrompt,
      skills: skills.map((s) => ({
        name: s.name,
        slug: s.slug,
        kind: s.kind,
        body: s.body,
      })),
    },
    task: task
      ? {
          id: task.id,
          name: task.name,
          description: task.description,
          status: task.status,
          approvalGate: task.approvalGate,
          attachments,
        }
      : null,
    goal,
    mcpConnections: input.mcpConnectionNames ?? agent.mcpConnectionIds,
    filesystemGrants: agent.filesystemGrants,
    repos: agent.repoAccess,
    environment,
    collaborationList: agent.collaborationList,
    inboxAccess: agent.inboxAccess,
    envNames,
    runnerPreference: agent.runnerPreference,
  };
}

/** Network wall check (§5.5): limited environments deny everything not allowlisted. */
export function networkAllowed(
  environment: SessionManifest["environment"],
  host: string,
): boolean {
  if (!environment || environment.networking === "open") return true;
  return environment.allowedHosts.some(
    (h) => h === host || host.endsWith("." + h),
  );
}

File: agentos/acl/grants.py
from __future__ import annotations
"""
Least-privilege ACL engine. Port of src/acl/grants.ts.
Default deny: everything an agent may touch must be listed in its manifest.
"""
from typing import Literal, Optional

from agentos.domain.types import FilesystemGrant, SessionManifest

FsOp = Literal["list", "read", "write", "mkdir", "delete"]


def normalize_path(raw: str) -> Optional[str]:
    """Normalize a virtual path; returns None when escaping the root."""
    parts = raw.replace("\\", "/").split("/")
    out: list[str] = []
    for p in parts:
        if p in ("", "."):
            continue
        if p == "..":
            return None  # traversal denied
        out.append(p)
    return "/" + "/".join(out)


def path_within(prefix: str, path: str) -> bool:
    p = normalize_path(prefix)
    q = normalize_path(path)
    if not p or not q:
        return False
    if p == "/":
        return True
    sep_p = p if p.endswith("/") else p + "/"
    return q == p or q.startswith(sep_p)


def grant_for(grants: list[FilesystemGrant], path: str) -> FilesystemGrant:
    """Combined grant for a path across all the agent's folder grants."""
    merged = FilesystemGrant(folderPath=path, canRead=False, canWrite=False, canDelete=False)
    for g in grants:
        if path_within(g.folderPath, path):
            merged.canRead = merged.canRead or g.canRead
            merged.canWrite = merged.canWrite or g.canWrite
            merged.canDelete = merged.canDelete or g.canDelete
    return merged


def check_fs_op(grants: list[FilesystemGrant], op: FsOp, path: str) -> tuple[bool, str]:
    """Returns (ok, reason). Deny unless the grant covers the op."""
    norm = normalize_path(path)
    if not norm:
        return False, f"path escape denied: {path}"
    g = grant_for(grants, norm)
    if op in ("list", "read") and not g.canRead:
        return False, f"no read grant for {norm}"
    if op in ("write", "mkdir") and not g.canWrite:
        return False, f"no write grant for {norm}"
    if op == "delete" and not g.canDelete:
        return False, f"no delete grant for {norm}"
    return True, ""


def network_allowed(environment: Optional[object], host: str) -> bool:
    """Network wall check (§5.5): limited environments deny everything not allowlisted."""
    if environment is None:
        return True
    networking = getattr(environment, "networking", "limited")
    if networking == "open":
        return True
    allowed_hosts: list[str] = getattr(environment, "allowedHosts", [])
    return any(h == host or host.endswith("." + h) for h in allowed_hosts)


def build_session_manifest(
    *,
    project_id: str,
    session_id: str,
    agent: object,
    skills: list[object],
    env_names: list[str],
    environment: Optional[object],
    task: Optional[object],
    goal: Optional[object],
    attachments: list[object],
    mcp_connection_names: Optional[list[str]] = None,
) -> dict:
    """Build a session manifest dict from ORM objects."""
    from agentos.domain.types import (
        SessionManifest,
        SessionManifestAgent,
        SessionManifestEnvironment,
        SessionManifestGoal,
        SessionManifestTask,
        TaskStatus,
    )

    a = agent
    skill_list = [
        {"name": s.name, "slug": s.slug, "kind": s.kind, "body": s.body}
        for s in skills
    ]
    env_block = (
        SessionManifestEnvironment(
            networking=environment.networking,
            allowedHosts=environment.allowed_hosts,
        )
        if environment
        else None
    )
    task_block = (
        SessionManifestTask(
            id=task.id,
            name=task.name,
            description=task.description,
            status=task.status,
            approvalGate=task.approval_gate,
            attachments=[att.model_dump() if hasattr(att, "model_dump") else att for att in attachments],
        )
        if task
        else None
    )
    goal_block = (
        SessionManifestGoal(
            id=goal.id,
            title=goal.title,
            spec=goal.spec,
            definitionOfDone=[d.text for d in goal.definition_of_done],
            progressLog=goal.progress_log,
        )
        if goal
        else None
    )
    return SessionManifest(
        sessionId=session_id,
        projectId=project_id,
        agent=SessionManifestAgent(
            id=a.id,
            name=a.name,
            title=a.title,
            model=a.model,
            foundationalPrompt=a.foundational_prompt,
            rolePrompt=a.role_prompt,
            skills=skill_list,
        ),
        task=task_block,
        goal=goal_block,
        mcpConnections=mcp_connection_names if mcp_connection_names is not None else a.mcp_connection_ids,
        filesystemGrants=a.filesystem_grants,
        repos=a.repo_access,
        environment=env_block,
        collaborationList=a.collaboration_list,
        inboxAccess=a.inbox_access,
        envNames=env_names,
        runnerPreference=a.runner_preference,
    ).model_dump()
